src/middlewared/middlewared/plugins/system/__init__.py

```diff
@@ -6,44 +6,47 @@
 from .utils import FIRST_INSTALL_SENTINEL, BOOTENV_FIRSTBOOT_SENTINEL, lifecycle_conf
 
 
-def firstboot(middleware):
+def init_routines(middleware):
+    try:
+        with open('/proc/sys/kernel/random/boot_id', 'r') as f:
+            lifecycle_conf.SYSTEM_BOOT_ID = f.read().strip()
+    except FileNotFoundError:
+        middleware.logger.error('Failed to read boot_id from /proc/sys/kernel/random/boot_id')
+        lifecycle_conf.SYSTEM_BOOT_ID = str(uuid.uuid4())
+
     if os.path.exists(BOOTREADY):
         lifecycle_conf.SYSTEM_READY = True
     elif os.path.exists(FIRST_INSTALL_SENTINEL):
         lifecycle_conf.SYSTEM_FIRST_BOOT = True
         # Delete sentinel file before making clone as we
         # we do not want the clone to have the file in it.
         os.unlink(FIRST_INSTALL_SENTINEL)
-
         if middleware.call_sync('system.is_enterprise'):
             config = middleware.call_sync('datastore.config', 'system.advanced')
             middleware.call_sync('datastore.update', 'system.advanced', config['id'], {'adv_autotune': True})
 
-
-def firstboot_after_upgrade(middleware):
     if not os.path.exists(BOOTENV_FIRSTBOOT_SENTINEL):
         os.makedirs(os.path.dirname(BOOTENV_FIRSTBOOT_SENTINEL), mode=0o700, exist_ok=True)
         with open(BOOTENV_FIRSTBOOT_SENTINEL, 'w'):
             pass
 
         lifecycle_conf.SYSTEM_BOOT_ENV_FIRST_BOOT = True
 
+    # upstream openZFS changed the name of
+    # these scripts
+    try:
+        os.symlink("/usr/bin/zarcstat", "/usr/sbin/arcstat")
+    except FileExistsError:
+        pass
 
-def read_system_boot_id(middleware):
     try:
-        with open('/proc/sys/kernel/random/boot_id', 'r') as f:
-            return f.read().strip()
-    except FileNotFoundError:
-        middleware.logger.error('Failed to read boot_id from /proc/sys/kernel/random/boot_id')
-        return str(uuid.uuid4())
+        os.symlink("/usr/bin/zarcsummary", "/usr/sbin/arc_summary")
```

Comment:
Won't these paths be RO?

Reply:
Maybe we should generate symlink when generating install iso?

Reply:
I was testing on a box where I removed all safeguards 😬 I'm going to close this and we'll fix by symlinking during build.